fix(ivy): fix multiple nested views removal from ViewContainerRef

waterplea opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #38201


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

waterplea commented:
@pkozlowski-opensource please help me out with testing this, I couldn't find where the units for this part are located at. Should I test is some other way? Also I might messed up commit message. I saw `ivy` scope on many commits but I see it's not in the list at the given commit messages link:
https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit

waterplea commented:
@googlebot I signed it!

AndrewKushnir commented:
Hi @waterplea, please find replies below.

> I couldn't find where the units for this part are located at. Should I test is some other way?

You can add tests to the `packages/core/test/acceptance/view_container_ref_spec.ts` file. There is a number of existing tests there that can be a good starting point for additional tests to verify the change in this PR.

> Also I might messed up commit message. I saw ivy scope on many commits but I see it's not in the list at the given commit messages

The `ivy` scope was deprecated since Ivy became the default rendering engine. Please use `core` instead (so your commit message would start with `fix(core) ...`).

Please let us know if you have any questions.

Thank you.

AndrewKushnir commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -27,7 +27,18 @@ import {isProceduralRenderer, RComment, RElement} from './interfaces/renderer';
+  nativeInsertBefore,
+  nativeNextSibling,
+  nativeParentNode,
+  removeView
```
Since the `removeView` function was removed from `node_manipulation.ts` file, it should be removed from the import as well (there is a [failing CI job](https://app.circleci.com/pipelines/github/angular/angular/19527/workflows/ab00ed0d-18e5-4175-9217-54142b91f9ad/jobs/779487) that indicates that).

waterplea replied:
Silly question, but why don't just add prettier and hook it to precommit?

AndrewKushnir replied:
Currently the `clang-format` is used and you can find more information on source code formatting [here](https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#formatting-your-source-code). I think looking into adding a precommit hook makes sense. Could you please create a new ticket in the "Issues" section of the repo, so this proposal is captured for further review?

pkozlowski-opensource commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -304,8 +315,9 @@ export function createContainerRef(
-        removeView(this._lContainer, adjustedIdx);
+        const detachedView = detachView(this._lContainer, adjustedIdx);
         removeFromArray(this._lContainer[VIEW_REFS]!, adjustedIdx);
+        detachedView && destroyLView(detachedView[TVIEW], detachedView);
```
could you please re-write it as an `if(...) {...}`  statement for readability?

pkozlowski-opensource reviewed (changes requested):
I'm marking it as "Request changes" since:
* CI is failing
* we are missing test(s)
* I've left a stylistic comment

Please let me know if the @AndrewKushnir comment unblocked you in terms of adding tests.

waterplea commented:
Couldn't get tests to run, I get 404 error when I run `test-ivy-aot`:
```
WARNING: Download from https://mirror.bazel.build/nodejs.org/dist/v12.14.1/node-v12.14.1-win-x64.zip fail
ed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET return
ed 404 Not Found
```
So I pushed tests with commented fix to make sure they fail. Then I will push fix to make sure they pass.

BTW, why not add lint-staged so all changes go through linters before they can be committed and pushed? This way issues like unused imports will be caught early. 

Once this PR is finished I will squash all commits with proper message.

AndrewKushnir commented:
Hi @waterplea,

> Couldn't get tests to run, I get 404 error when I run test-ivy-aot:

Could you please let us know which command do you run? For the acceptance tests that you added, the following command can be used:

```
bazel test --config=ivy //packages/core/test/acceptance:acceptance 
```

Note: the error that is thrown in the `test_ivy_aot` [CircleCI job](https://app.circleci.com/pipelines/github/angular/angular/19615/workflows/2be1060d-db5e-48e1-820f-8342cf58025d/jobs/780815) indicates that the test should probably moved to a different (or a new one) `describe` block, since there is a call to `inject` in the `beforeEach` function that belongs to the parent `describe`.

> BTW, why not add lint-staged so all changes go through linters before they can be committed and pushed? This way issues like unused imports will be caught early.

See my reply here: https://github.com/angular/angular/pull/38317#discussion_r465378150. Could you please create a new ticket in the "Issues" section and describe this proposal?

> Once this PR is finished I will squash all commits with proper message.

Please let us know if you have any questions.

Thank you.

waterplea commented:
@AndrewKushnir Thank you, that command was failing on Windows with `Action failed to execute process.cc(199): CreateProcessW` but I got it running on Mac.

I've moved test into separate describe and tried it to make sure it was failing before the fix.

waterplea commented:
@AndrewKushnir @pkozlowski-opensource I think I got it all done this time, but builds are failing with something unrelated:
```
error An unexpected error occurred: "https://registry.yarnpkg.com/@bazel/worker/-/worker-0.37.0.tgz: Request failed \"404 Not Found\"".
```

AndrewKushnir commented:
@waterplea, thanks for the update. It looks like the CI error is unrelated to your PR and is likely caused by the recent CircleCI problem (see [here](https://status.circleci.com/)). Could you please try to rebase this PR once again? This should trigger CI to rerun. Thank you.

waterplea commented:
@AndrewKushnir any idea what this could be about? I've just rebased onto master because CI told me something has changed over there and now I'm getting this:
![image](https://user-images.githubusercontent.com/11832552/89618414-b57f4e00-d894-11ea-9424-384bae6535d7.png)

AndrewKushnir commented:
Hi @waterplea, it looks like it was a problem on CircleCI side. I've restarted the CI and the `setup` job (that failed before) worked correctly. Please let us know if you notice this problem again. Thank you.

waterplea commented:
Hi @AndrewKushnir thanks for a quick response. Last failing CI job is also failing for an unclear reason:
`/bin/bash: ./.circleci/env.sh: No such file or directory`
Rerun is greyed out for me, could you assist please?

JoostK commented:
> Hi @AndrewKushnir thanks for a quick response. Last failing CI job is also failing for an unclear reason:
> `/bin/bash: ./.circleci/env.sh: No such file or directory`
> Rerun is greyed out for me, could you assist please?

You appear to have a lot of bad luck with CircleCI, it's not nearly this broken for me. Anyway, I restarted the failing job.

waterplea commented:
@pkozlowski-opensource looks like it's all done and CI jobs all passed. Could you take another look please?

AndrewKushnir commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -304,8 +304,12 @@ export function createContainerRef(
+        const detachedView = detachView(this._lContainer, adjustedIdx);
+
+        if (detachedView) {
+          removeFromArray(this._lContainer[VIEW_REFS]!, adjustedIdx);
```
Thanks for the fix @waterplea 👍 

I think it'd be great to add a comment here to explain why the calls are executed in this order, similar to what was described in https://github.com/angular/angular/issues/38201#issuecomment-664402562.

mhevery reviewed (approved):
Please take care of @AndrewKushnir comment, otherwise LGTM.

AndrewKushnir commented:
FYI, I've started [tests in Google's codebase](http://test/OCL:326292478:BASE:326292528:1597262056415:1b7b4c72) and will update this thread once I have the results.

@pkozlowski-opensource could you please have a look when you get a chance to see if you have additional feedback?

Thank you.

waterplea commented:
Added the comment. Thanks everybody, I didn't do much other than follow your directions, but now I at least know my way around a little :) This should make following contributions easier shall there be some.

pkozlowski-opensource reviewed (approved):
LGTM overall but left some comments that I would love to see addressed.
Also, the commit message should link to the issue being fixed, something like `Fixes #38201`

Adding cleanup label so the above can be taken care of

waterplea commented on packages/core/test/acceptance/view_container_ref_spec.ts:
```diff
@@ -936,6 +936,65 @@ describe('ViewContainerRef', () => {
+        declarations: [AppComponent, TemplateDirective],
+      });
+
+      const fixture: ComponentFixture<AppComponent> = TestBed.createComponent(AppComponent);
```
Looks like my TypeScript service was acting up, because it didn't show me code complete suggestions, that's why I added it.

AndrewKushnir commented:
Hi @waterplea, just wanted to mention that tests in Google's codebase passed and all approvals were received, so we should be able to add this PR to the merge queue.

One quick comment: it looks like commit message will require a small update: you've added the `Fixes` part to the commit header, but we usually put it at the very end of the commit body. Could you please update it? So the commit message may look like this (I've also included some additional information based on your comment in the code, feel free to change it as you see fit):

```
fix(core): fix multiple nested views removal from ViewContainerRef

When removal of one view causes removal of another one from the same
ViewContainerRef it triggers an error with views length calculation. This commit
fixes this bug by removing a view from the list of available views before invoking
actual view removal (which might be recursive and relies on the length of the list
of available views).

Fixes #38201.
```

Thank you.

waterplea committed:
fix(core): fix multiple nested views removal from ViewContainerRef

When removal of one view causes removal of another one from the same
ViewContainerRef it triggers an error with views length calculation. This commit
fixes this bug by removing a view from the list of available views before invoking
actual view removal (which might be recursive and relies on the length of the list
of available views).

Fixes #38201.

waterplea commented:
Hey @AndrewKushnir that's great, thank you. I've changed the commit message and re-pushed.

AndrewKushnir commented:
Thanks for the update @waterplea. I've added the "merge" label, so this PR will be ready for merge once CI is completed successfully.
